pkg/tests/schema_pulumi_test.go

```diff
@@ -4159,3 +4012,3298 @@ func TestMakeTerraformResultNilVsEmptyMap(t *testing.T) {
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+    ~ prov:index/test:Test: (update)
+        [id=newid]
+        [urn=urn:pulumi:test::test::prov:index/test:Test::mainRes]
+      ~ tests: [
+          + [1]: {
+                  + nested    : "val5"
+                }
+          ~ [2]: {
+                  ~ nested: "val3" => "val6"
+                }
+          - [3]: {
+                  - nested: "val4"
+                }
+        ]
+Resources:
+    ~ 1 to update
+    1 unchanged
+`),
+			map[string]interface{}{
+				"tests[1]":        map[string]interface{}{"kind": "ADD_REPLACE"},
+				"tests[2].nested": map[string]interface{}{"kind": "UPDATE_REPLACE"},
+				"tests[3]":        map[string]interface{}{"kind": "DELETE_REPLACE"},
+			},
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+    +-prov:index/test:Test: (replace)
+        [id=newid]
+        [urn=urn:pulumi:test::test::prov:index/test:Test::mainRes]
+      ~ tests: [
+          + [1]: {
+                  + nested    : "val5"
+                }
+          ~ [2]: {
+                  ~ nested: "val3" => "val6"
+                }
+          - [3]: {
+                  - nested: "val4"
+                }
+        ]
+Resources:
+    +-1 to replace
+    1 unchanged
+`),
+		},
+		{
+			"two added and two removed shuffled, no overlaps",
+			[]string{"val1", "val2", "val3", "val4"},
+			[]string{"val5", "val6", "val1", "val2"},
+			map[string]interface{}{
+				"tests[0]": map[string]interface{}{},
+				"tests[1]": map[string]interface{}{},
+				"tests[2]": map[string]interface{}{"kind": "DELETE"},
+				"tests[3]": map[string]interface{}{"kind": "DELETE"},
+			},
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+    ~ prov:index/test:Test: (update)
+        [id=newid]
+        [urn=urn:pulumi:test::test::prov:index/test:Test::mainRes]
+      ~ tests: [
+          + [0]: "val5"
+          + [1]: "val6"
+          - [2]: "val3"
+          - [3]: "val4"
+        ]
+Resources:
+    ~ 1 to update
+    1 unchanged
+`),
+			map[string]interface{}{
+				"tests[0]": map[string]interface{}{"kind": "ADD_REPLACE"},
+				"tests[1]": map[string]interface{}{"kind": "ADD_REPLACE"},
+				"tests[2]": map[string]interface{}{"kind": "DELETE_REPLACE"},
+				"tests[3]": map[string]interface{}{"kind": "DELETE_REPLACE"},
+			},
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+    +-prov:index/test:Test: (replace)
+        [id=newid]
+        [urn=urn:pulumi:test::test::prov:index/test:Test::mainRes]
+      ~ tests: [
+          + [0]: "val5"
+          + [1]: "val6"
+          - [2]: "val3"
+          - [3]: "val4"
+        ]
+Resources:
+    +-1 to replace
+    1 unchanged
+`),
+			map[string]interface{}{
```

Comment:
Did you use `go generate` to make these test cases? :) There are a lot.

Reply:
I wish, the various behaviours are inconsistent enough to not be abstractable. I've grouped them as much as possible